src/main/java/org/ice4j/stack/StunStack.java

```diff
@@ -1639,4 +1522,117 @@ public static void logPacketToPcap(
         }
     }
 
+    /**
+     * Class which performs periodic collection of expired transactions.
+     * It's execution is controlled outside by {@link #schedule()}
+     * and {@link #cancel()} methods. Whenever expired transactions collector
+     * is scheduled it does self reschedule with fixed delay
+     * of StunServerTransaction.LIFETIME_MILLIS,
+     * until {@link #serverTransactions} is empty, in that case it self-cancel
+     * further execution and need to be scheduled again when new item added
+     * to {@link #serverTransactions} container.
+     */
+    private final class ExpiredServerTransactionsCollector
+    {
+        /**
+         * Runnable which walks {@link #serverTransactions}, check if
+         * transaction is expired and if so - remove it
+         * from {@link #serverTransactions}.
+         * Self-cancels when {@link #serverTransactions} is empty
+         */
+        private final Runnable collector = new Runnable()
+        {
+            @Override
+            public void run()
+            {
+                try
+                {
+                    synchronized (serverTransactions)
+                    {
+                        final int transactionsBeforeCollection
+                            = serverTransactions.size();
+
+                        long now = System.currentTimeMillis();
+
+                        for (Iterator<StunServerTransaction> i
+                                    = serverTransactions.values().iterator();
+                                i.hasNext();)
+                        {
+                            StunServerTransaction serverTransaction = i.next();
+
+                            if (serverTransaction == null)
+                            {
+                                i.remove();
+                            }
+                            else if (serverTransaction.isExpired(now))
+                            {
+                                i.remove();
+                                serverTransaction.expire();
+                            }
+                        }
+
+                        logger.fine("Non-expired server transactions "
+                            + "count " + serverTransactions.size()
+                            + ", transactions before collection "
+                            + transactionsBeforeCollection);
+
+                        if (serverTransactions.isEmpty())
```

Comment:
i wonder if, rather than cancelling the job from within itself, `schedule` should check for `serverTransactions.isEmpty()` and just not re-schedule it? i don't see the value in this cancelling itself, since by definition it will be done right after this anyway, no?

Reply:
@bbaldino  This runnable is scheduled with fixed delay, so it is automatically re-scheduled after specified interval time. This cancellation is a “mimic” of old behavior when thread is self cancel when there’re is nothing to expire. I could remove it since it is much cheaper now when there is no extra thread is paid for each StunStack object.

Reply:
Ah oops, ok I missed that.  Understood.